File: smartthings_tools/my_devices/media.py
from __future__ import annotations

from typing import Any

from ..base import BaseTool, ToolResult
from ..schemas.my_devices import MediaPlaybackInput, MediaVolumeInput


class MediaPlaybackTool(BaseTool):
    name = "media_playback"
    description = (
        "미디어 재생을 제어합니다. 재생, 일시정지, 정지, 빨리감기, 되감기를 지원합니다."
    )
    args_schema = MediaPlaybackInput

    def execute(self, *, device_id: str, action: str, **kwargs: Any) -> ToolResult:
        data = self.client.command(device_id, "mediaPlayback", action)
        if "error" in data:
            return ToolResult(
                success=False, error=data.get("error", {}).get("message", str(data))
            )
        return ToolResult(success=True, data={"device_id": device_id, "action": action})


class MediaVolumeTool(BaseTool):
    name = "media_volume"
    description = "미디어 디바이스(스피커 등)의 볼륨을 조절합니다."
    args_schema = MediaVolumeInput

    def execute(
        self,
        *,
        device_id: str,
        action: str,
        volume: int | None = None,
        **kwargs: Any,
    ) -> ToolResult:
        if action == "set":
            if volume is None:
                return ToolResult(
                    success=False, error="volume is required when action is 'set'."
                )
            data = self.client.command(device_id, "audioVolume", "setVolume", [volume])
        elif action == "up":
            data = self.client.command(device_id, "audioVolume", "volumeUp")
        elif action == "down":
            data = self.client.command(device_id, "audioVolume", "volumeDown")
        else:
            return ToolResult(
                success=False, error=f"Unsupported volume action: {action}"
            )

        if "error" in data:
            return ToolResult(
                success=False, error=data.get("error", {}).get("message", str(data))
            )
        return ToolResult(
            success=True,
            data={"device_id": device_id, "action": action, "volume": volume},
        )

File: smartthings_tools_rs/src/my_devices/media.rs
use schemars::schema::RootSchema;
use schemars::schema_for;
use serde_json::Value;

use crate::client::SmartThingsClient;
use crate::common_tools::{extract_error_msg, has_error};
use crate::schemas::my_devices::*;
use crate::tool::{Tool, ToolResult};

pub struct MediaPlaybackTool<'a> {
    pub client: &'a SmartThingsClient,
}

impl Tool for MediaPlaybackTool<'_> {
    fn name(&self) -> &'static str {
        "media_playback"
    }
    fn description(&self) -> &'static str {
        "미디어 재생을 제어합니다. 재생, 일시정지, 정지, 빨리감기, 되감기를 지원합니다."
    }
    fn parameters_schema(&self) -> RootSchema {
        schema_for!(MediaPlaybackInput)
    }

    fn execute(&self, args: Value) -> ToolResult {
        let input: MediaPlaybackInput = match serde_json::from_value(args) {
            Ok(v) => v,
            Err(e) => return ToolResult::fail(format!("ValidationError: {e}")),
        };
        let data =
            self.client
                .command(&input.device_id, "mediaPlayback", &input.action, None, None);
        if has_error(&data) {
            return ToolResult::fail(extract_error_msg(&data));
        }
        ToolResult::ok(serde_json::json!({ "device_id": input.device_id, "action": input.action }))
    }
}

pub struct MediaVolumeTool<'a> {
    pub client: &'a SmartThingsClient,
}

impl Tool for MediaVolumeTool<'_> {
    fn name(&self) -> &'static str {
        "media_volume"
    }
    fn description(&self) -> &'static str {
        "미디어 디바이스(스피커 등)의 볼륨을 조절합니다."
    }
    fn parameters_schema(&self) -> RootSchema {
        schema_for!(MediaVolumeInput)
    }

    fn execute(&self, args: Value) -> ToolResult {
        let input: MediaVolumeInput = match serde_json::from_value(args) {
            Ok(v) => v,
            Err(e) => return ToolResult::fail(format!("ValidationError: {e}")),
        };
        let data = match input.action.as_str() {
            "set" => {
                let vol = match input.volume {
                    Some(v) => v,
                    None => return ToolResult::fail("volume is required when action is 'set'."),
                };
                self.client.command(
                    &input.device_id,
                    "audioVolume",
                    "setVolume",
                    Some(vec![Value::from(vol)]),
                    None,
                )
            }
            "up" => self
                .client
                .command(&input.device_id, "audioVolume", "volumeUp", None, None),
            "down" => {
                self.client
                    .command(&input.device_id, "audioVolume", "volumeDown", None, None)
            }
            other => return ToolResult::fail(format!("Unsupported volume action: {other}")),
        };
        if has_error(&data) {
            return ToolResult::fail(extract_error_msg(&data));
        }
        ToolResult::ok(
            serde_json::json!({ "device_id": input.device_id, "action": input.action, "volume": input.volume }),
        )
    }
}
